feat: malicious token screening on transactions

imblue-dabadee opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

Introduces token screening on incoming tokens received in transactions. This comes in the form of two different alerts (yellow and red).

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: Added an alert if an incoming token is malicious or suspicious.

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: Token screening on incoming tokens in transactions
  Scenario: user initiates a transaction where they receive malicious tokens
    Given the user is on the `Transaction request` screen
    And they are receiving tokens that is flagged as malicious

    When user views the screen
    Then they will see a red alert on `You receive`
    And a red `Review alert` button
    
  Scenario: user initiates a transaction where they receive suspicious tokens
    Given the user is on the `Transaction request` screen
    And they are receiving tokens that is flagged as suspicious

    When user views the screen
    Then they will see a yellow alert on `You receive`
```

- For a malicious token, you can swap for `0x69e8b9528cabda89fe846c67675b5d73d463a916` on a swap website.
- For a suspicious token, you can swap for `0xd0cd466b34a24fcb2f87676278af2005ca8a78c4` on a swap website.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->
**1. No yellow alert**

<img width="402" height="853" alt="Screenshot 2025-11-18 at 12 00 50 PM" src="https://github.com/user-attachments/assets/c176193c-26ff-44eb-9b10-160381da3356" />

**2. No red alert**

<img width="398" height="843" alt="Screenshot 2025-11-18 at 12 01 34 PM" src="https://github.com/user-attachments/assets/148f9c4e-6f5c-4d79-b0d2-11932c39dbfb" />

### **After**

<!-- [screenshots/recordings] -->
**1. Yellow alert**

<img width="386" height="841" alt="Screenshot 2025-11-18 at 11 41 51 AM" src="https://github.com/user-attachments/assets/4028a106-357a-42fd-a7e9-2b60301b1185" />

**2. Red alert**
<img width="383" height="618" alt="Screenshot 2025-11-19 at 9 15 25 AM" src="https://github.com/user-attachments/assets/01e04dfe-5abc-42fa-84f6-dca1bf05477c" />

**When you click the inline alert**
<img width="403" height="667" alt="Screenshot 2025-11-19 at 9 15 40 AM" src="https://github.com/user-attachments/assets/1ddeed60-64e7-4c2e-8305-3fecca11e715" />

**When you click the 'Review alert' button**
<img width="389" height="641" alt="Screenshot 2025-11-19 at 9 17 41 AM" src="https://github.com/user-attachments/assets/9c305ec1-534e-49a6-bc8c-6ea39a4d28cc" />


## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.




<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds malicious/suspicious token screening to transaction confirmations with inline alerts, selectors, strings, and tests.
> 
> - **Confirmations UI/UX**:
>   - Show `AlertRow` on `BalanceChangeRow` label when `hasIncomingTokens` is true, with style override in `alert-row.styles`.
>   - Compute `hasIncomingTokens` in `BalanceChangeList` and pass to rows.
> - **Alerts System**:
>   - New `useTokenTrustSignalAlerts` hook to derive alerts from token scan results; integrated into `useConfirmationAlerts`.
>   - Added `RowAlertKey.IncomingTokens` and new alert keys `TokenTrustSignalMalicious`/`TokenTrustSignalWarning` with metrics mappings.
> - **State Selectors**:
>   - New `selectMultipleTokenScanResults` to read `PhishingController.tokenScanCache` for multiple tokens.
> - **Localization**:
>   - English strings for malicious/suspicious token alerts.
> - **Dependencies**:
>   - Upgrade `@metamask/phishing-controller` to `^16.1.0`.
> - **Tests**:
>   - Unit tests for `BalanceChangeRow`, `AlertRow`, `useTokenTrustSignalAlerts`, `useConfirmationAlerts`, and phishing selectors.
> - **Test fixtures**:
>   - Update initial background state to include `addressScanCache` and `tokenScanCache`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit ff7ae25421cd55a871411b09dc71cb062a48fbf6. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

imblue-dabadee committed:
en copy

imblue-dabadee committed:
basic alert

imblue-dabadee committed:
set up rowalertkey

imblue-dabadee committed:
setup AlertKeys

imblue-dabadee committed:
integrate TokenTrustSignalAlerts into confirmationAlerts

imblue-dabadee committed:
feat: selector

imblue-dabadee committed:
fix: en.json text key

imblue-dabadee committed:
feat: useTokenTrustSignalAlert

imblue-dabadee committed:
feat: add warning alert key

imblue-dabadee committed:
feat: ui integration

imblue-dabadee committed:
fix: make chainId required for selector

imblue-dabadee committed:
tests

imblue-dabadee committed:
Merge remote-tracking branch 'origin/main' into feat/malicious-token-screening-on-transactions

imblue-dabadee committed:
fix: bump phishing-controller version

imblue-dabadee committed:
Merge branch 'main' into feat/malicious-token-screening-on-transactions

imblue-dabadee committed:
fix: add missing metric names

imblue-dabadee committed:
chore: yarn dedupe

imblue-dabadee committed:
feat: multiple token scan result selector

imblue-dabadee committed:
refactor: use selectMultipleTokenScanResults

imblue-dabadee committed:
tests

imblue-dabadee committed:
Merge branch 'feat/malicious-token-screening-on-transactions' of https://github.com/MetaMask/metamask-mobile into feat/malicious-token-screening-on-transactions

imblue-dabadee committed:
Merge branch 'main' into feat/malicious-token-screening-on-transactions

imblue-dabadee committed:
fix: lint

AugmentedMode commented on app/components/UI/SimulationDetails/BalanceChangeRow/BalanceChangeRow.tsx:
```diff
@@ -32,23 +34,41 @@ const BalanceChangeRow: React.FC<BalanceChangeRowProperties> = ({
+    }
+    if (hasIncomingTokens) {
+      return (
+        <AlertRow
```
Do you not need a confirmationID to be passed in here ?

imblue-dabadee committed:
chore: remove commented lines

imblue-dabadee commented on locales/languages/en.json:
```diff
@@ -100,6 +100,16 @@
+    },
+    "token_trust_signal": {
+      "malicious": {
+        "title": "Malicious token",
```
Ok will do - that was the copy provided by @mindofmar so not sure if that went through some specific approval or not

imblue-dabadee committed:
chore: update copy

imblue-dabadee committed:
Merge branch 'main' into feat/malicious-token-screening-on-transactions

imblue-dabadee committed:
Revert "chore: update copy"

This reverts commit cab282b9981d8879acfb9c4710ada08e39e2e08e.

imblue-dabadee committed:
fix: css

imblue-dabadee committed:
Merge branch 'feat/malicious-token-screening-on-transactions' of https://github.com/MetaMask/metamask-mobile into feat/malicious-token-screening-on-transactions

imblue-dabadee committed:
fix: revert removing the text color to outgoing tokens

imblue-dabadee committed:
Merge branch 'main' into feat/malicious-token-screening-on-transactions

imblue-dabadee committed:
refactor: simplify the phishingController code

imblue-dabadee committed:
chore: remove unused function

imblue-dabadee committed:
refactor: simplify selectMulitpleTokenScanResults

Commit:
fix: increase Node.js heap size for iOS bundle generation in CI

The js-bundle-size-check job was experiencing JavaScript heap out of memory errors during the iOS bundle generation step. This adds NODE_OPTIONS with --max_old_space_size=8192 (8GB) to prevent these failures, following the same pattern used in other memory-intensive CI jobs.

imblue-dabadee committed:
feat: bump phishign controller

imblue-dabadee committed:
refactor: use reduce

imblue-dabadee committed:
Merge branch 'main' into feat/malicious-token-screening-on-transactions

imblue-dabadee committed:
fix: add addressScanCache to background state

AugmentedMode reviewed (dismissed):
LGTM, QAd and everything is working as expected

OGPoyraz commented on app/components/UI/SimulationDetails/BalanceChangeList/BalanceChangeList.tsx:
```diff
@@ -28,6 +29,12 @@ const BalanceChangeList: React.FC<BalanceChangeListProperties> = ({
 
+  const hasIncomingTokens = useMemo(
+    () =>
+      balanceChanges.some((balanceChange) => balanceChange.amount.isPositive()),
```
Do we want to skip out positive native incoming ones?

OGPoyraz replied:
Do we want to skip out positive native incoming ones as they probably won't be a malicious?

OGPoyraz commented on app/components/Views/confirmations/hooks/alerts/useTokenTrustSignalAlerts.ts:
```diff
@@ -0,0 +1,94 @@
+        title,
+        severity,
+        isBlocking: false,
+      });
```
What happens if there are multiple incoming tokens one malicious and one warning? Not sure multiple alerts on same `RowAlertKey` will work out.

Maybe we can collect all warnings and malicious token symbols and show it at once in the alert modal?

imblue-dabadee committed:
fix: return only the highest severity alert

imblue-dabadee committed:
test: only one alert and highest severity

imblue-dabadee committed:
Merge branch 'main' into feat/malicious-token-screening-on-transactions

imblue-dabadee commented on locales/languages/en.json:
```diff
@@ -104,6 +104,16 @@
+    "token_trust_signal": {
+      "malicious": {
+        "title": "Malicious token",
+        "message": "This token has been identified as malicious. Interacting with this token may result in a loss of funds."
```
Only returning one alert now so no need to update the copy. Thank you!

AugmentedMode reviewed (approved):
LGTM, nice changes
